operators/knative-istio-controller/src/charm.py

```diff
@@ -62,32 +65,68 @@ def render(self):
             verbs=None,
         )
 
-        # FIXME: remove hardcoded knative-serving 
+        # FIXME: remove hardcoded knative-serving
         args = {
-            "gateway_name": "knative-custom-gateway",
-            "gateway_namespace": self.model.name,
             "name": "istio-controller",
             "namespace": self.model.name,
-            "knative_serving": "knative-serving"
+            "knative_serving": "knative-serving",
         }
 
         templates = [
-            env.get_template("deployment.yaml.j2").render(**args),
-            env.get_template("config.yaml.j2").render(**args),
-            env.get_template("rbac.yaml.j2").render(**args),
+            self.env.get_template("deployment.yaml.j2").render(**args),
+            self.render_config_istio(args),
+            self.env.get_template("rbac.yaml.j2").render(**args),
             # TODO: Temporarily removed as I try to get other gateways to work
             # env.get_template("gateway.yaml.j2").render(**args),
-            env.get_template("peer_auth.yaml.j2").render(**args),
+            self.env.get_template("peer_auth.yaml.j2").render(**args),
         ]
 
         return codecs.load_all_yaml("\n---\n".join(templates))
 
+    def render_config_istio(self, args):
+        istio_pilot_app = self.model.get_app("istio-pilot")
+        gateway = self.model.relations["gateway"]
+
+        if len(gateway) == 0:
+            raise CheckFailedError("Missing required relation for gateway", BlockedStatus)
+
+        try:
+            data = gateway[0].data[istio_pilot_app]
+
+            args = {
+                **args,
+                "gateway_name": data["gateway-name"],
+                "gateway_namespace": data["gateway-namespace"],
+            }
+            return self.env.get_template("config.yaml.j2").render(**args)
+        except Exception as error:
+            self.log.error(
+                "Encountered the following error when parsing gateway relation", f"{str(error)}"
+            )
+            raise CheckFailedError(
+                "Unexpected error when parsing gateway relation data. See log", BlockedStatus
+            )
```

Comment:
@DomFleischmann was mentioning today how the functionality here of extracting a gateway name and namespace from the `istio:gateway` relation would be better as a library in the istio-pilot charm rather than as code here.  Similar to how the prometheus/grafana integrations work.  Then this charm does not need to know the details of how istio populates the data in the relation, it just uses the functionality provided by the library